### PR TITLE
Phoenix Connector: Handle parallel execution of scans explicitly in Trino

### LINKED ISSUE #10143: Queries to the Phoenix connector can use up all of the TrinoServer's heap memory
Phoenix has the notion of guidepost, equal width stats, in order to gauge how to parallelize a query.

For Trino this means that Phoenix will parallelize queries even for a single split when the Phoenix plan suggests to use the RoundRobinIterator.
What's makes this more interesting is that each Phoenix query segment (Scan) will prefetch some data (2MB worth of rows by default, with all overhead it could 3-5mb on the heap).

The effect: I loaded a decent amount of data (about 800gb uncompressed, about 150bg compressed) onto a very small (single node, because I wanted to do some debugging) Phoenix and Trino cluster. This is a simple count(*) query, and since Phoenix does not support aggregate pushdown all rows are actually loaded by the TrinoServer. What happened is that TrinoServer quickly ran out of heap.

When I looked at the heapdump I noticed for each split Phoenix created many thousands of scans per split, and each scan started pre-fetching. In effect the TrinoServer now had many thousands of scans open, each of which started fetching, filling the heap with millions of HBase Result objects and their associated byte[]. 10000's of scans using some 3-5mb of heap adds up quickly.

One could argue that this is a contrived case where we use an aggregate that cannot be pushed with a lot of data on a very small custer.

Possible solutions:
1. Do nothing and assume that this is not a common case. This is not entirely unreasonable. The default guidepost width in 300mb, which I had lowered to 10MB and then 100MB. Even at 300mb, this would create ~2700 scans for my dataset, and use up 8-12gb of heap without even starting executing the split.
2. Never allow Phoenix to parallelize inside a Split (i.e. always use a ConcatIterator instead of the RoundRobinIterator), and leave all parallelize to Trino. Small queries might be slower now, since we're losing out of the pre-fetching.
3. Put in some logic that allows the RoundRobinIterator for a Split when we have less than N scans. N should be low, like 10 or so, definitely < 100. So small queries would still be fast.
4. Make this configurable with a boolean enable/disable intra-split-parallelism (default off!)
5. Make it configurable with the maximum parallelism should be (i.e. 3. above but configurable, default 10). Not that we can tell the RoundRobinIterators how many scans to prefetch, so it always all or nothing.

I'd lean towards 2. or 3. and not introduce more config options.

It's also not at all clear how to isolate this into test.

### PR DESCRIPTION
**Pull request by lhofhansl**
Fixes #10143

Remove Phoenix intra-split parallelism, which sometimes "uncontrollably" uses a lot of the heap, and generate more fine grained splits in Trino instead.

Why 20 scans per split? There is no right value. 20 is a good compromise. When Guideposts are enabled (default in "on" Phoenix with a width of 300mb), each split represents at most 6gb worth of scanned data.

This *is* configurable in Phoenix, but currently it does not make much sense in Phoenix to make the guideposts width larger, or reduce it below 100mb. So 20 is about right. Any number between 10 and 60 will do. (With 20gb and 300mb guideposts it does not make much sense to set this higher then 60).

When guideposts are disabled, there is one scan per region and this change does not affect anything.

I'd prefer this hardcoded to a reasonable value instead of added yet another config option.

### CODE REVIEW

**Comment by lhofhansl** [comment]
Test seems unrelated.

I did some more manual:
1. Set Phoenix's guidepost width as low as 10MB, nobody would set this as low, as the Phoenix client by default would not be able to handle that (the execution threadpool is way too small and needs to configured). That led to 3801 splits in Trino. The query finished (albeit about 40% slower). Without this PR the query would use up all memory and fail and make the Trinoserver unresponsive.
2. Left Phoenix's guidepost width at default 300MB. That led to a nice 153 splits, and the query finished without using much memory as it should be.

So this should be good. In the future when Phoenix gets smarter about guidepost handling we can address this again.

**Comment by lhofhansl** [comment]
Use Guava Lists.partition rather than my braindead way of doing it before.

**Comment by lhofhansl** [comment]
Question: Does Trino shuffle Splits? If not, I might want to shuffle them here so that all splits for a region are run together, thus hotspotting that region reserver.

**Comment by lhofhansl** [comment]
Ping... Should I add someone else to the review?
Querying larger datasets (as in size per worker), this is a serious issue.

**Review by vincentpoon — APPROVED** [review]
lgtm @lhofhansl

**Comment by lhofhansl** [comment]
Had some time today...

1. Made it configurable after all. This allow for controlling parallelism from the Trino side alone. The default remains 20.
2. Rebased onto lastest master.
3. Shuffles the list of splits now, so that each worker will hit region server randomly per split.

@hashhar Wanna have a look?

What's a good way to unit-test this?

**Comment by lhofhansl** [comment]
I'll fix the test soon,

**Comment by lhofhansl** [comment]
Ping.

**Comment by hashhar** [comment]
Sorry for the delay @lhofhansl. I'm taking a look now.

File: plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixSplitManager.java [hunk 1]
```diff
@@ -58,6 +59,7 @@
         implements ConnectorSplitManager
 {
     private static final Logger log = Logger.get(PhoenixSplitManager.class);
+    private static final int MAX_SCANS_PER_SPLIT = 20;
```
Comment (hashhar) [inline plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixSplitManager.java]: Justification for this value in code comment (and commit description) would be helpful.

IIUC this mechanism reduces the number of scans being performed on Phoenix by grouping some of them within a single split?

File: docs/src/main/sphinx/connector/phoenix.rst [hunk 2]
```diff
@@ -56,6 +56,10 @@ Property Name                                      Required   Description
                                                               default the location is ``/hbase``
 ``phoenix.config.resources``                       No         Comma-separated list of configuration files (e.g. ``hbase-site.xml``) to use for
                                                               connection properties.  These files must exist on the machines running Trino.
+``phoenix.max-scans-per-split``                    No         Maximum number of HBase scans allowed to be grouped into a single split. Default is 20.
+                                                              Lower values will lead to more splits in Trino.
+                                                              Can also be set via session propery ``phoenix.max_scans_per_split``.
```
Comment (hashhar) [inline docs/src/main/sphinx/connector/phoenix.rst:61]: Remove the `phoenix` prefix since that'll depend on actual catalog name.

File: docs/src/main/sphinx/connector/phoenix.rst [hunk 3]
```diff
@@ -56,6 +56,10 @@ Property Name                                      Required   Description
...
+                                                              (This setting has no effect when guideposts are disabled in Phoenix.)
```
Comment (hashhar) [inline docs/src/main/sphinx/connector/phoenix.rst:62]: Is this something that a Phoenix user will most likely know or should we link to something on Phoenix side?

File: plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixConfig.java [hunk 4]
```diff
@@ -16,16 +16,22 @@
...
 
+import javax.validation.constraints.Max;
+import javax.validation.constraints.Min;
 import javax.validation.constraints.NotNull;
 
 import java.util.List;
 
 public class PhoenixConfig
 {
+    static final int MAX_ALLOWED_SCANS_PER_SPLIT = 1000;
```
Comment (hashhar) [inline plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixConfig.java:30]: Add some code comment (and commit description) with regards to significance of this value.

File: plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixSplitManager.java [hunk 5]
```diff
@@ -154,10 +155,11 @@ public ConnectorSplitSource getSplits(
                  * Each split is handled by a single ConcatResultIterator
                  * (See PhoenixClient.getResultSet(...))
                  */
-                for (List<Scan> splitScans : Lists.partition(scans, MAX_SCANS_PER_SPLIT)) {
+                for (List<Scan> splitScans : Lists.partition(scans, maxScansPerSplit)) {
                     inputSplits.add(new PhoenixInputSplit(splitScans, regionSize, regionLocation));
                 }
             }
+            Collections.shuffle(inputSplits);
```
Comment (hashhar) [inline plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixSplitManager.java:162]: Why is this needed, not obvious to me.

File: plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixConfig.java [hunk 6]
```diff
@@ -52,4 +58,19 @@ public PhoenixConfig setResourceConfigFiles(String files)
...
+
+    @Min(1)
+    @Max(MAX_ALLOWED_SCANS_PER_SPLIT)
+    public int getMaxScansPerSplit()
+    {
+        return maxScansPerSplit;
+    }
+
+    @Config("phoenix.max-scans-per-split")
+    @ConfigDescription("The maximum number of HBase scans that Phoenix will allow per split.")
```
Comment (hashhar) [inline plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixConfig.java:70]: The description feels incorrect since Phoenix doesn't have any control I'm guessing.

Maybe change to `Maximum number of HBase scans that can be performed in a single split.`

File: docs/src/main/sphinx/connector/phoenix.rst [hunk 7]
```diff
@@ -56,6 +56,10 @@ Property Name                                      Required   Description
                                                               default the location is ``/hbase``
 ``phoenix.config.resources``                       No         Comma-separated list of configuration files (e.g. ``hbase-site.xml``) to use for
                                                               connection properties.  These files must exist on the machines running Trino.
+``phoenix.max-scans-per-split``                    No         Maximum number of HBase scans allowed to be grouped into a single split. Default is 20.
```
Comment (hashhar) [inline docs/src/main/sphinx/connector/phoenix.rst:59]: WDYT about - `Maximum number of HBase scans that can be performed in a single split`.

**Review by hashhar — COMMENTED** [review]
Still trying to wrap my head around this - some initial questions.

**Comment by hashhar** [comment]
Since you've added a session property I guess you can add a functional test that with a non-default value of max-scans-per-split queries work and return expected results.

I'm not sure if there's some way to verify that Phoenix indeed served fewer scans.

File: plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixSplitManager.java (hunk 1 above)
Comment (lhofhansl) [reply plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixSplitManager.java]: Phoenix uses two metrics to split queries into multiple HBase scans:
1. HBase regions. Each region is handled by a separate (set of) scan(s)
2. Since HBase regions can get large, Phoenix keeps additional metrics: Guideposts, where it records the value of a PK (which are ordered) every N bytes (default 300MB). And if enabled each Guideposts that falls into the range of keys touched by a query turns into a separate scan.

So with this patch we find a compromise. A scan per region leaves compute resources unused, and a scan per Guidepost is overkill especially for larger tables.

I'll put something in the comment.

File: docs/src/main/sphinx/connector/phoenix.rst (hunk 2 above)
Comment (lhofhansl) [reply docs/src/main/sphinx/connector/phoenix.rst:61]: Arrgh. Yes you are right of course.

File: docs/src/main/sphinx/connector/phoenix.rst (hunk 3 above)
Comment (lhofhansl) [reply docs/src/main/sphinx/connector/phoenix.rst:62]: Good call... I'll link here: https://phoenix.apache.org/update_statistics.html

File: plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixSplitManager.java (hunk 5 above)
Comment (lhofhansl) [reply plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixSplitManager.java:162]: We now may get multiple Splits for the same region, and it seems Trino runs them in order per worker. (Right?)
We want to avoid a worker running multiple splits against the same region concurrently while splits for other regions are waiting in the queue.

File: plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixConfig.java (hunk 6 above)
Comment (lhofhansl) [reply plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixConfig.java:70]: Agreed.

File: plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixSplitManager.java (hunk 5 above)
Comment (hashhar) [reply plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixSplitManager.java:162]: Add this as a comment. cc: @sopel39 to confirm split execution ordering, I'm not very familiar with the scheduler off the top of my head.

**Comment by lhofhansl** [comment]
In the end HBase will have served the exact same number of scans, what we are controlling is the parallelism, and most importantly avoiding two level scheduling where Trino schedules its tasks and within each task Phoenix schedules its chunks (i.e. guildepost size scans). Lemme think on a test. Semantically the output will always be the same.

File: docs/src/main/sphinx/connector/phoenix.rst (hunk 2 above)
Comment (lhofhansl) [reply docs/src/main/sphinx/connector/phoenix.rst:61]: Should I rename this to "max_hbase_scans_per_split" or something? So that the general "max_scans_per_split" is available to other connectors if needed?

**Comment by hashhar** [comment]
If the split count is controllable using this config then you may be able to assert on the query stats.

Looks good overall.

File: docs/src/main/sphinx/connector/phoenix.rst (hunk 2 above)
Comment (hashhar) [reply docs/src/main/sphinx/connector/phoenix.rst:61]: I don't think that should be an issue. Session properties are already "namespaced". I think existing convention suggests to keep existing name.

**Comment by lhofhansl** [comment]
The trick is controlling the guideposts in Phoenix in a test. They are updated asynchronously only. I can probably force a major compaction, wait for it to finish, and then count the guide posts.

File: plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixConfig.java (hunk 4 above)
Comment (lhofhansl) [reply plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixConfig.java:30]: In theory no upper limit is needed. When this is set to a very large value Trino is going to have one split per HBase region.
I'll add a comment.

**Comment by hashhar** [comment]
I don't feel very strongly about a test for this since it'll most likely end up flaky and the execution before and after the change is the same (and is enabled by default so already "tested").

**Comment by lhofhansl** [comment]
Added comments, squashed, and rebased.

**Review by hashhar — APPROVED** [review]

File: plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixConfig.java [hunk 8]
```diff
@@ -16,17 +16,35 @@
...
+import javax.validation.constraints.Min;
 import javax.validation.constraints.NotNull;
 
 import java.util.List;
 
 public class PhoenixConfig
 {
+    /*
+     * Normally this should be kept between 1 and 70 or so (with HBase and Phoenix defaults).
+     * A large value here makes sense when the Guidepost-width in Phoenix has been reduced.
```
Comment (hashhar) [inline plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixConfig.java:32]: IMO remove this comment and instead add `A large value here makes sense when the Guidepost-width in Phoenix has been reduced.` to the comment for `maxScansPerSplit`.

File: plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixConfig.java [hunk 9]
```diff
@@ -52,4 +70,19 @@ public PhoenixConfig setResourceConfigFiles(String files)
...
+
+    @Min(1)
+    @Max(MAX_ALLOWED_SCANS_PER_SPLIT)
+    public int getMaxScansPerSplit()
+    {
+        return maxScansPerSplit;
+    }
+
+    @Config("phoenix.max-scans-per-split")
+    @ConfigDescription("Maximum number of HBase scans that can be performed in a single split.")
```
Comment (hashhar) [inline plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixConfig.java:82]:
```suggestion
    @ConfigDescription("Maximum number of HBase scans that will be performed in a single split.")
```

**Review by hashhar — APPROVED** [review]
LGTM % editorials.

File: plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixSplitManager.java [hunk 10]
```diff
@@ -147,8 +150,17 @@ public ConnectorSplitSource getSplits(
...
+                /*
+                 * Handle parallel execution explicitly in Trino rather than internally in Phoenix.
+                 * Each split is handled by a single ConcatResultIterator
+                 * (See PhoenixClient.getResultSet(...))
+                 */
+                for (List<Scan> splitScans : Lists.partition(scans, maxScansPerSplit)) {
+                    inputSplits.add(new PhoenixInputSplit(splitScans, regionSize, regionLocation));
+                }
             }
+            // shuffle to distribute load across HBase regions
```
Comment (sopel39) [inline plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixSplitManager.java:162]: The order of splits in list shouldn't make a difference. See `io.trino.execution.scheduler.SplitPlacementPolicy#computeAssignments`. Splits are `Set` there

Comment (lhofhansl) [reply plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixSplitManager.java:162]: Thanks. So the shuffle is unneeded, I'll remove it.

**Comment by lhofhansl** [comment]
Rebased. Removed the shuffle. Adjusted comments.

**Review by hashhar — APPROVED** [review]
